django/contrib/auth/forms.py

```diff
@@ -212,14 +219,32 @@ def clean(self):
         password = self.cleaned_data.get('password')
 
         if username is not None and password:
+            self.pre_authenticate(username)
             self.user_cache = authenticate(self.request, username=username, password=password)
             if self.user_cache is None:
+                # Cache failed login with delay.
+                cache.set(
+                    '%s:%s' % (self.FAILED_LOGIN_CACHE_PREFIX, sha256(force_bytes(username)).hexdigest()),
+                    timezone.now() + timedelta(seconds=self.DELAY_AFTER_FAILED_LOGIN),
+                    60
+                )
                 raise self.get_invalid_login_error()
             else:
                 self.confirm_login_allowed(self.user_cache)
 
         return self.cleaned_data
 
+    def pre_authenticate(self, username):
+        # Check for failed logins delay in the cache.
+        auth_delay = cache.get('%s:%s' % (
+            self.FAILED_LOGIN_CACHE_PREFIX,
+            sha256(force_bytes(username)).hexdigest()
+        ))
+        if auth_delay:
+            wait_for = (auth_delay - timezone.now()).total_seconds()
+            if wait_for > 0:
+                sleep(wait_for)
```

Comment:
Won't this result in denial-of-service attacks?

Reply:
A form validation error seems sensible to me

Reply:
So, quite right. The other issue is that an attacker could still cause a denial-of-service to specific users where their username is known.

[`django-axes`](https://pypi.org/project/django-axes/) has a disabled by default option [`AXES_LOCK_OUT_BY_COMBINATION_USER_AND_IP`](https://django-axes.readthedocs.io/en/latest/4_configuration.html#configuring-project-settings) which comes in handy from the other direction where a user may be blocked by IP due to another user on the same shared connection, e.g. in the same office, getting their password wrong too many times.

We could either just always do the `user+ip` combo, or we could have a setting to select between `disabled`, `user`, or `user+ip`. Some people may want to disable this functionality, but I think that we should default to `user+ip` which would enable this, but give some leniency. `user` will probably be too aggressive if there is no way to turn it off, even if it gives better protection against distributed attacks.

Do we also want to allow for a configurable number of attempts? Just a single value, or so many per-user and per-ip within a certain timeframe?

Reply:
A different idea might be to add a new exception type, e.g. `TooManyRequests`, and add a branch to https://github.com/django/django/blob/f4e93919e4608cfc50849a1f764fd856e0917401/django/core/handlers/exception.py#L54 returning a 429 HTTP status code. Of course one could argue that Django shouldn't add an exception for each status code (I agree) but rate limiting may be important enough given it's OWASP classification.

All implementations using `time.sleep()` will suffer from the same DoS problem. Also, `time.sleep()` will probably have to be replaced when more of Django is reworked with ASGI in mind.